color picker: pass in pipe (not piece) via picker ready signal

dtorop opened the pull request:
We previously received `piece`, and from that derived the current `pipe`. Instead just pass the current `pipe` via `DT_SIGNAL_CONTROL_PICKERDATA_READY`.

In one case we derived the current `module` from the passed-in `piece`, but as we are already sending `module` via the signal, just use that directly.

Besides it being repetitious to pass in `piece`, the danger is that in the interim between the signal being raised and processed, `piece` could no longer be a valid pointer due to the pixelpipe being recreated. As we currently trigger recreates every time data is received, via `DT_DEV_PIPE_REMOVE`, this is a very real risk.

This may fix #14812. But note that there's more work to be done, as we shouldn't need to recreate the entire pixelpipe and wipe the cache every time we receive color picker data. At the very least this slows down work when editing with the per-iop colorpicker active.

Also note that colorpicker always runs in preview pipe, hence it's repetitious (if future-looking) to even pass around the current `pipe` value.

TurboGit commented:
@dtorop : I'm not sure this makes a difference. See for example 47a9f0dc469952f7efba0612ebd6786387245b88. This fixes and issue by using `piece` which is stable and safe to use during the whole process.

Here at the end it seems it is equivalent, so maybe a pure cosmetic change, no? Or do you have evidence that it fixes an issue? @jenshannoschwalm knows quite well the pipe circuitry, maybe I'm missing something.

dtorop commented:
@TurboGit: Thanks for looking this over! 

> @dtorop : I'm not sure this makes a difference. See for example [47a9f0d](https://github.com/darktable-org/darktable/commit/47a9f0dc469952f7efba0612ebd6786387245b88). This fixes and issue by using `piece` which is stable and safe to use during the whole process.

I believe that 47a9f0dc469952f7efba0612ebd6786387245b88 is different. That commit works because `piece` is stable during a given run of a pixelpipe. But we can't guarantee `piece` is stable between two different runs of the pipe. Particularly if `DT_DEV_PIPE_REMOVE` has triggered a recreation of the pipe.

The latter is what this PR tries to fix: `DT_SIGNAL_CONTROL_PICKERDATA_READY` is raised during one run of the pipe, but by the time it is handled in the GUI thread, that pipe run may be complete and the nodes in it cleaned up. In which case `piece` is no longer valid. If we pass `pipe` with the signal, we can still determine what we actually need to know -- the current `pipe`, which will be stable between runs. 
 
> Here at the end it seems it is equivalent, so maybe a pure cosmetic change, no?

Not cosmetic, due to averting a potential crash! But cosmetic in the sense that except for that, the functionality should be equivalent.

> Or do you have evidence that it fixes an issue?

Alas, no. I worked backwards from a single backtrace, to try to figure out what the cause of the crash could be. But wasn't able to replicate the crash again (either with or without this PR's commits applied). So this is all surmise from that backtrace and examination of the code. I think this is a rare-to-see crash.

> @jenshannoschwalm knows quite well the pipe circuitry, maybe I'm missing something.

Would be most interested for @jenshannoschwalm thoughts. It's a few weeks ago that I paid attention to this, I wish I had a better memory of its details.

jenshannoschwalm commented:
> I believe that [47a9f0d](https://github.com/darktable-org/darktable/commit/47a9f0dc469952f7efba0612ebd6786387245b88) is different. That commit works because `piece` is stable during a given run of a pixelpipe. But we can't guarantee `piece` is stable between two different runs of the pipe. Particularly if `DT_DEV_PIPE_REMOVE` has triggered a recreation of the pipe.

I think this analysis is correct.

I am currently investigating https://github.com/darktable-org/darktable/blob/d5455e7c24ee574fbe6b12cfe1f67a1c932497ed/src/views/darkroom.c#L407

which uses data for proxies and gui_expose of modules. In certain situations it is called twice for a single change in darkroom (i see this while retouch is "in-focus"). Don't know what that implies though ...

dtorop commented:
> I am currently investigating
> 
> https://github.com/darktable-org/darktable/blob/d5455e7c24ee574fbe6b12cfe1f67a1c932497ed/src/views/darkroom.c#L407
>
> which uses data for proxies and gui_expose of modules. In certain situations it is called twice for a single change in darkroom (i see this while retouch is "in-focus"). Don't know what that implies though ...

@jenshannoschwalm: Interesting... I may be biased by my (slow-going) work widget-ifying print view, but I suspect that the architecture of massive `expose()` callbacks is getting brittle. The expose function needs to integrate so many things, at this point (masks, color picker, image from one of two pipes, ISO 12646 borders, loading message, guide lines) not to mention the potential `gui_post_expose()` overlays from duplicate and snapshots.

I wonder if a better architecture will need to be a variety of widgets overlaid, to do each of these things, each with their own rules about when they update. Updates then would generally happen when a signal was received (e.g. pipe completed, or colorpicker sample is ready).

But that may not be the problem you're seeing...

jenshannoschwalm commented:
Currently we just run the pipe job if required in a certain order. Maybe we should that in a while condition until "all jobs are done" and only then do all "proxy work"?

jenshannoschwalm commented:
Also not sure how "early exit of a pipe" contributes to this.

dtorop commented:
@TurboGit: I tried to use github to resolve a merge conflict with base branch. The resulting commit a0cb740e1789eac1ba3ddade83e21f740aba1112 looks really ugly to me! If this PR goes through, I can simply rebase to make this more sensible?

TurboGit commented:
> I can simply rebase to make this more sensible?

Yes please!

dtorop commented:
> > I can simply rebase to make this more sensible?
> 
> Yes please!

@TurboGit: Rebased and force pushed this as a single (cleaner!) commit. I updated the commit message to note that this appears to fix #14929 as well.

TurboGit reviewed (approved):
Thanks!

TurboGit reviewed (approved):
Manually merged with conflict resolution. Thanks!
